streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -741,6 +757,23 @@ public void close(boolean clean,
      * @param records   the records
      */
     public void addRecords(final TopicPartition partition, final Iterable<ConsumerRecord<byte[], byte[]>> records) {
+        // if condition put here in case of restarts and rebalances to check for correct timestamp
+        if (recordInfo.queue() != null && getPartitionTime(partition) == RecordQueue.UNKNOWN) {
+            final OffsetAndMetadata metadata;
+            if (!eosEnabled) {
```

Comment:
I am not an expert on this API, but I would expect that even if we commit using the producer, the consumer should still be able to read the metadata.

Did you verify that we cannot retrieve the metadata if EOS is enabled? If this is the case, I would claim it's a bug that need to be fix.

Reply:
Alright, I will check on that.

Reply:
Ok, so after some checks, I have discovered the following: the producer and consumer are not subscribed to the same TopicPartition (however, that is not necessarily the reason for the failure I'm about to describe.) I'm not 100% sure if this was supposed to happen. When producer committed the transaction, that is supposed to mean the consumer group coordinator was notified to commit the offsets right? However, when the consumer available to StreamTask is called (i.e. ```committed(TopicPartition)```), null is returned (indicating no offsets were committed). 

Producer's sendOffsetsToTransaction requires the user to also enter a consumer group id (which is specified by user config) which we send the offsets to commit to. It is possible that unless the user specifies the right consumer group id, we would not be able to retrieve associated metadata.

Reply:
> the producer and consumer are not assigned to the same TopicPartition

Not sure what you mean by this? A producer is never _assigned_ any partitions. We use the term _assigned_ for consumers only. Can you clarify?

> It is possible that unless the user specifies the right consumer group id, we would not be able to retrieve associated metadata.

Absolutely. The `group.id` used to commit offsets via the producer must match the `group.id` used by the consumer to read those offsets. In Kafka Streams the `application.id` is used as `group.id` for the consumer and it's also passed into `producer.sendOffsetsToTransaction()`.

Hence, I still think that we don't need to handle EOS differently to get the partition-time.

Does this help?

Reply:
Ah, forgot that producer isn't assigned partitions. 

Let me see what I can do to handle to the EOS case then.